man/dhcpd-pools.1.in
.TH DHCPD-POOLS "1" "2017-11-15" "@VERSION@" "User Commands"
.SH NAME
dhcpd-pools \- ISC dhcpd pools usage analysis
.SH SYNOPSIS
.SY dhcpd-pools
.OP \-\-config file
.OP \-\-leases file
.OP \-\-sort nimcptTe
.OP \-\-reverse
.OP \-\-format tHcxXjJ
.OP \-\-mustach template
.OP \-\-output file
.OP \-\-limit nr
.OP \-\-color when
.OP \-\-warning percent
.OP \-\-critical percent
.OP \-\-warn\-count number
.OP \-\-crit\-count number
.OP \-\-snet\-alarms
.OP \-\-minsize size
.OP \-\-perfdata
.OP \-\-version
.OP \-\-help
.YS
.SH DESCRIPTION
The program analyses ISC dhcpd shared network and pool usage and outputs the
results in a format selected by user.
.SH OUTPUT FIELDS
.TP
.I "shared net name"
Name of the shared-network for the range.
.TP
.I "first ip"
First IP in lease pool range.
.TP
.I "last ip"
Last IP in lease pool range.
.TP
.I "max"
Number of IPs which exist in a pool, shared network or all together.
.TP
.I "cur"
Number of leases currently in use.
.TP
.I "percent"
Percent of IPs currently in use compared to max.
.TP
.I "touch"
Number of IP's which appear in the lease file, but who's leases have expired.
A touched IP is either expired or abandoned.  The touched IP count is
somewhat misleading when you try to determine if an IP pool is big enough; it
is a better indicator of whether a pool is too large.
.TP
.I "t+c"
The sum of Touched and Currently in-use leases.
.TP
.I "t+c perc"
Percent of IPs either touched or currently in use, compared to max.
.TP
.I "bu"
Failover pair can allocate these addresses.  The count appears only if there
is failover configuration.
.TP
.I "bu perc"
Percent of addresses that failover pair can allocate.  The percent appears
only if there is failover configuration.
.SH OPTIONS
.TP
\fB\-c\fR, \fB\-\-config\fR=\fIFILE\fR
Path to the dhcpd.conf file.  If the dhcpd.conf has include files they
can be analysed separately, that can be useful when trying to understand
or monitor subset of data.
.TP
\fB\-l\fR, \fB\-\-leases\fR=\fIFILE\fR
Path to the dhcpd.leases file.
.TP
\fB\-s\fR, \fB\-\-sort\fR=\fI[nimcptTe]\fR
Sort ranges by chosen fields as a sorting keys.  Keys weight from left to
right, i.e., if more weighting keys are equal next one is used.  The IP
field is default sort key.
.TP
\fB\-r\fR, \fB\-\-reverse\fR
Sort results in reverse order.
.TP
\fB\-f\fR, \fB\-\-format\fR=\fI[tHcxXjJ]\fR
Output format.
Text
.RI ( t ).
Full-html
.RI ( H )
page output.  In html page critical and warning thresholds can be visualized
with \-\-color=always option.  The
.RI ( c )
stands for comma-separated values.  Output format xml
.RI ( x )
is similar to the dhcpstatus Perl module output.  The extended xml
.RI ( X )
format will print ethernet address details.  The
.RI ( j )
will output in json format, which can be extended with
.RI ( J )
to include ethernet address.
.IP
The default format is
.IR @OUTPUT_FORMAT@ .
.TP
\fB\-\-mustach\fR=\fITEMPLATE\fR
Output using mustach
.I template
file.  This is useful when the native output formats controlled with
.B \-\-format
option do not provide what you need.  See below example mustach template
that is using all available {{tags}} to demonstrate what can be displayed
and how.
.IP
@bindir@/dhcpd-pools --config @docdir@/dhcpd.conf --leases
@docdir@/dhcpd.leases --mustach @docdir@/mustach.template
.TP
\fB\-o\fR, \fB\-\-output\fR=\fIFILE\fR
.I File
where output is written.  Default is stdout.
.TP
\fB\-L\fR, \fB\-\-limit\fR=\fINR\fR
The
.I NR
will limit what will be printed.  Syntax is similar to
.IR chmod (1)
permission string.  The
.I NR
limit string uses two digits which vary between
.IR 0 \ to \ 7 .
The first digit determines which headers to display, and the second digit
determines which numeric analysis tables to include in the output.  The
following values are "OR'd" together to create the desired output.  The
default is
.IR @OUTPUT_LIMIT@ .
.IP
.TS
tab(:);
ll.
0\fI1\fR:Print ranges
0\fI2\fR:Print shared networks
0\fI4\fR:Print total summary
\fI1\fR0:Print range header
\fI2\fR0:Print shared network header
\fI4\fR0:Print total summary header
.TE
.IP
The output limit for total summary has special meaning in
.B \-\-warning
and
.B \-\-critical
alarming context.  When the alarming is in use, and total is not wanted
to be seen then in the case of alarming returning success nothing is
printed.
.TP
\fB\-\-color\fR=\fIwhen\fR
Use yellow for warning, red for critical, green for suppressed by \-\-minsize
and blue when \-\-snet\-alarms is the cause of supression or shared network
does not have any ranges.  The
.I when
string can be
.BR always ,
.BR never ,
or
.BR auto .
Default is auto, that uses colors when command is running in interactive
terminal.  With use of
.B \-\-warning
or
.B \-\-critical
coloring thresholds can be changed, but one must also use
.B \-\-format=text
to avoid turning on alarting mode.
.TP
\fB\-\-skip\fR=\fIwhen\fR
The
.I when
can be one of the following:
.IR ok ,
.IR warning ,
.IR critical ,
.IR minsize ,
or
.IR suppressed .
The skipping criterias are exact match with colors in \-\-color option.
.TP
\fB\-\-warning\fR=\fIpercent\fR
Turn on alarm output format, and specify percentage number which will
cause an alarm.  If either a range or shared network will exceed
warning level return value of the command is
.BR 1 .
If only range monitoring is needed one can use limit option for scoping,
for example
.IR \-L10 .
To monitor shared network only the limit would be
.IR \-L20 .
If warning percentage is not specified it defaults to
.BR @ALARM_WARN@ .
The
.I percent
argument allows fractions, e.g., 88.8, to be used.
.TP
\fB\-\-critical\fR=\fIpercent\fR
The option is similar to warning, with exception of return value which
is
.BR 2 .
If critical percentage is not specified it defaults to
.BR @ALARM_CRIT@ .
.TP
\fB\-\-warn\-count\fR=\fInumber\fR
A
.I number
of free leases before alarm is raised.  When specified both \-\-warning
.I percent
and count
.I number
are required to be exceeded in order to alarm criteria being fulfilled.
.IP
This option is intended to be used in setup where very large and small
shared-networks and ranges co-exists.  In such environments percent based
alarming can lead to either flood of alarms about small ranges, or way too
great overhead of free addresses in large shared-networks.  Suggested usage
is to set percentage to a level that makes small ranges to ring, and set the
count to match level when an enormous shared-network is too few free leases.
.IP
Defaults to 2^32, that is size of entire IPv4 address space.
.TP
\fB\-\-crit\-count\fR=\fInumber\fR
Same as \-\-warn\-count, but for critical alarms.
.TP
\fB\-\-snet\-alarms
Suppress range alarms that are part of shared networks.  Use of this option
will keep alarm criteria applied to ranges that are not part of shared-net
along with shared-net alarms.  This option may help reducing alarm noise for
configurations that has lots of small ranges in big shared-networks.
.TP
\fB\-\-minsize\fR=\fIsize\fR
Ignore ranges and shared networks that are smaller or equal to the
defined size.  This option is meaningful only in context of alarming, and
will intended to suppress for example single host ranges.  By default this
option is not in use.
.TP
\fB\-p\fR, \fB\-\-perfdata\fR
Print additional performance data, like lease count, touched leases and
backup leases.  This option is meaningful only in context of alarming and
will print lots of data, if there are many networks.  By default this option
is not in use.
.TP
\fB\-A\fR, \fB\-\-all\-as\-shared\fR
Treat all stand-alone subnets as shared-network with named formed from it's
CIDR.  By default this option is not in use for backwards compatibility.
.TP
\fB\-\-ip\-version\fR=\fI4|6\fR
Force command to read configuration and leases files in IPv4 or IPv6 mode.
Notice that when inputs do not match with what is forced analysis output is
garbage.  This option should not be necessary to use, and exists only to
allow debugging.
.TP
\fB\-v\fR, \fB\-\-version\fR
Print version information to standard output and exit successfully.
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help to standard output and exit successfully.
.SH EXAMPLES
.TP
Print ranges header, and analysis.
$ dhcpd-pools \-L 11 \-c dhcpd.conf \-l dhcpd.leases
.br
Ranges:
.br
shared net name [...]
.TP
Print shared networks and totals, both headers and results
$ dhcpd-pools \-L 66 \-c dhcpd.conf \-l dhcpd.leases shared net name
.br
[...]
.TP
Alarming
$ dhcpd-pools \-c dhcpd.conf \-l dhcpd.leases \-\-critical 80.1 \-\-warning 75
.br
CRITICAL: dhcpd-pools: Ranges; crit: 14 warn: 22 ok: 220 Shared nets; crit: 1 warn: 0 ok: 4
.IP
$ dhcpd-pools \-c dhcpd.conf \-l dhcpd.leases \-L 22 \-\-critical 70 \-\-warning 50
.br
[no-output]
.br
Suppress printing OK, and make alarm only to go off if shared networks
exceed critial or warning levels.
.SH FILES
.TP
@DHCPDCONF_FILE@
ISC dhcpd configuration file.
.TP
@DHCPDLEASE_FILE@
ISC dhcpd lease file.
.TP
@docdir@/prometheus.template
Prometheus text file collector mustach template.
.SH AUTHORS
Original design by Sami Kerola.
.br
uthash by Troy D. Hanson.
.br
XML support by Dominic Germain, Sogetel inc.
.br
IPv6 support by Cheer Xiao.
.br
Mustache templating support by José Bollo.
.SH LICENSE
The dhcpd-pools uses FreeBSD License, uthash uses BSD license, the mustache
uses Apache License, and the gnulib modules are mostly, but not entirely,
GPL.
.SH "REPORTING BUGS"
Report bugs to
.MT @PACKAGE_BUGREPORT@
@PACKAGE_MAINTAINER@
.ME
.br
.UR @PACKAGE_URL@
Home page
.UE
.SH "SEE ALSO"
.BR dhcpd.leases (5),
.BR dhcpd.conf (5),
.BR chmod (1),
.UR https://mustache.github.io/
.UE


src/dhcpd-pools.c
/*
 * The dhcpd-pools has BSD 2-clause license which also known as "Simplified
 * BSD License" or "FreeBSD License".
 *
 * Copyright 2006- Sami Kerola. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *    1. Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *
 *    2. Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the
 *       distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR AND CONTRIBUTORS OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * The views and conclusions contained in the software and documentation are
 * those of the authors and should not be interpreted as representing
 * official policies, either expressed or implied, of Sami Kerola.
 */

/*! \file dhcpd-pools.c
 * \brief The main(), and core initialization.
 */

#include <config.h>

#include <errno.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include <stdio.h>
#include <limits.h>

#include "close-stream.h"
#include "closeout.h"
#include "error.h"
#include "progname.h"
#include "quote.h"
#include "xalloc.h"

#include "dhcpd-pools.h"

/* Function pointers */
int (*parse_ipaddr) (struct conf_t *state, const char *restrict src, union ipaddr_t *restrict dst);
void (*copy_ipaddr) (union ipaddr_t *restrict dst, const union ipaddr_t *restrict src);
const char *(*ntop_ipaddr) (const union ipaddr_t *ip);
double (*get_range_size) (const struct range_t *r);
int (*xstrstr) (struct conf_t *state, const char *restrict str);
int (*ipcomp) (const union ipaddr_t *restrict a, const union ipaddr_t *restrict b);
int (*leasecomp) (const struct leases_t *restrict a, const struct leases_t *restrict b);
void (*add_lease) (struct conf_t *state, union ipaddr_t *ip, enum ltype type);
struct leases_t *(*find_lease) (struct conf_t *state, union ipaddr_t *ip);

/*! \brief An option argument parser to populate state header_limit and
 * number_limit values.
 */
static int return_limit(const char c)
{
	if ('0' <= c && c < '8')
		return c - '0';
	error(EXIT_FAILURE, 0, "return_limit: output mask %s is illegal", quote(optarg));
	return 0;
}

/*! \brief Run time initialization. Global allocations, counter
 * initializations, etc are here. */
static void prepare_memory(struct conf_t *state)
{
	state->ranges = xmalloc(sizeof(struct range_t) * state->ranges_size);
	/* First shared network entry is all networks */
	state->shared_net_root = xcalloc(sizeof(struct shared_network_t), 1);
	state->shared_net_root->name = xstrdup("All networks");
	state->shared_net_head = state->shared_net_root;
}

/*! \brief The --skip option argument parser. */
static void skip_arg_parse(struct conf_t *state, char *optarg)
{
	enum {
		OPT_ARG_OK = 0,
		OPT_ARG_WARNING,
		OPT_ARG_CRITICAL,
		OPT_ARG_MINSIZE,
		OPT_ARG_SUPRESSED
	};

	char *const tokens[] = {
		[OPT_ARG_OK] = "ok",
		[OPT_ARG_WARNING] = "warning",
		[OPT_ARG_CRITICAL] = "critical",
		[OPT_ARG_MINSIZE] = "minsize",
		[OPT_ARG_SUPRESSED] = "suppressed",
		NULL
	};
	char *value;

	while (*optarg != '\0') {
		switch (getsubopt(&optarg, tokens, &value)) {
		case OPT_ARG_OK:
			state->skip_ok = 1;
			break;
		case OPT_ARG_WARNING:
			state->skip_warning = 1;
			break;
		case OPT_ARG_CRITICAL:
			state->skip_critical = 1;
			break;
		case OPT_ARG_MINSIZE:
			state->skip_minsize = 1;
			break;
		case OPT_ARG_SUPRESSED:
			state->skip_suppressed = 1;
			break;
		default:
			error(EXIT_FAILURE, 0, "unknown --skip specifier: %s", value);
		}
	}
}

/*! \brief Command line options parser. */
static char parse_command_line_opts(struct conf_t *state, int argc, char **argv)
{
	enum {
		OPT_SNET_ALARMS = CHAR_MAX + 1,
		OPT_WARN,
		OPT_CRIT,
		OPT_MINSIZE,
		OPT_WARN_COUNT,
		OPT_CRIT_COUNT,
		OPT_COLOR,
		OPT_SKIP,
		OPT_SET_IPV,
		OPT_MUSTACH
	};

	static struct option const long_options[] = {
		{"config", required_argument, NULL, 'c'},
		{"leases", required_argument, NULL, 'l'},
		{"color", required_argument, NULL, OPT_COLOR},
		{"skip", required_argument, NULL, OPT_SKIP},
		{"format", required_argument, NULL, 'f'},
		{"sort", required_argument, NULL, 's'},
		{"reverse", no_argument, NULL, 'r'},
		{"output", required_argument, NULL, 'o'},
		{"limit", required_argument, NULL, 'L'},
		{"mustach", required_argument, NULL, OPT_MUSTACH},
		{"version", no_argument, NULL, 'v'},
		{"help", no_argument, NULL, 'h'},
		{"snet-alarms", no_argument, NULL, OPT_SNET_ALARMS},
		{"warning", required_argument, NULL, OPT_WARN},
		{"critical", required_argument, NULL, OPT_CRIT},
		{"warn-count", required_argument, NULL, OPT_WARN_COUNT},
		{"crit-count", required_argument, NULL, OPT_CRIT_COUNT},
		{"minsize", required_argument, NULL, OPT_MINSIZE},
		{"perfdata", no_argument, NULL, 'p'},
		{"all-as-shared", no_argument, NULL, 'A'},
		{"ip-version", required_argument, NULL, OPT_SET_IPV},
		{NULL, 0, NULL, 0}
	};
	char output_format = '\0';
	int alarming = 0;

	while (1) {
		int c;

		c = getopt_long(argc, argv, "c:l:f:o:s:rL:pAvh", long_options, NULL);
		if (c == EOF)
			break;
		switch (c) {
		case 'c':
			/* config file */
			state->dhcpdconf_file = optarg;
			break;
		case 'l':
			/* lease file */
			state->dhcpdlease_file = optarg;
			break;
		case 'f':
			/* Output format */
			output_format = optarg[0];
			break;
		case 's':
			{
				/* Output sorting option */
				struct output_sort *p = state->sorts;
				size_t len;

				while (p && p->next)
					p = p->next;
				for (len = 0; len < strlen(optarg); len++) {
					if (state->sorts == NULL) {
						state->sorts =
						    xcalloc(1, sizeof(struct output_sort));
						p = state->sorts;
					} else {
						p->next = xcalloc(1, sizeof(struct output_sort));
						p = p->next;
					}
					p->func = field_selector(optarg[len]);
				}
			}
			break;
		case 'r':
			/* What ever sort in reverse order */
			state->reverse_order = 1;
			break;
		case 'o':
			/* Output file */
			state->output_file = optarg;
			break;
		case 'L':
			/* Specification what will be printed */
			state->header_limit = return_limit(optarg[0]);
			state->number_limit = return_limit(optarg[1]);
			break;
		case OPT_MUSTACH:
#ifdef BUILD_MUSTACH
			state->mustach_template = optarg;
			output_format = 'm';
#else
			error(EXIT_FAILURE, 0, "compiled without mustach support");
#endif
			break;
		case OPT_COLOR:
			state->color_mode = parse_color_mode(optarg);
			if (state->color_mode == color_unknown)
				error(EXIT_FAILURE, errno, "unknown color mode: %s", quote(optarg));
			break;
		case OPT_SKIP:
			skip_arg_parse(state, optarg);
			break;
		case OPT_SNET_ALARMS:
			state->snet_alarms = 1;
			break;
		case OPT_WARN:
			alarming = 1;
			state->warning = strtod_or_err(optarg, "illegal argument");
			break;
		case OPT_CRIT:
			alarming = 1;
			state->critical = strtod_or_err(optarg, "illegal argument");
			break;
		case OPT_WARN_COUNT:
			alarming = 1;
			state->warn_count = strtod_or_err(optarg, "illegal argument");
			break;
		case OPT_CRIT_COUNT:
			alarming = 1;
			state->crit_count = strtod_or_err(optarg, "illegal argument");
			break;
		case OPT_MINSIZE:
			state->minsize = strtod_or_err(optarg, "illegal argument");
			break;
		case OPT_SET_IPV:
			switch (optarg[0]) {
			case '4':
				set_ipv_functions(state, IPv4);
				break;
			case '6':
				set_ipv_functions(state, IPv6);
				break;
			default:
				error(EXIT_FAILURE, 0, "unknown --ip-version argument: %s", optarg);
			}
			break;
		case 'p':
			/* Print additional performance data in alarming mode */
			state->perfdata = 1;
			break;
		case 'A':
			/* Treat single networks as shared with network CIDR as name */
			state->all_as_shared = 1;
			break;
		case 'v':
			/* Print version */
			print_version();
		case 'h':
			/* Print help */
			usage(EXIT_SUCCESS);
		default:
			error(EXIT_FAILURE, 0, "Try %s --help for more information.", program_name);
		}
	}

	/* Use default dhcpd.conf when user did not define anything. */
	if (state->dhcpdconf_file == NULL)
		state->dhcpdconf_file = DHCPDCONF_FILE;
	/* Use default dhcpd.leases when user did not define anything. */
	if (state->dhcpdlease_file == NULL)
		state->dhcpdlease_file = DHCPDLEASE_FILE;
	/* Use default limits when user did not define anything. */
	if (state->header_limit == 8) {
		char const *default_limit = OUTPUT_LIMIT;

		state->header_limit = return_limit(default_limit[0]);
		state->number_limit = return_limit(default_limit[1]);
	}
	/* Output format is not defined, if alarm thresholds are then it's alarming, else use the
	 * default.  */
	if (output_format == '\0') {
		if (alarming == 1)
			output_format = 'a';
		else {
			const char *const default_format = OUTPUT_FORMAT;

			output_format = default_format[0];
		}
	}
	return output_format;
}

/*!\brief Start of execution.  This will mostly call other functions one
 * after another.
 *
 * \return Return value indicates success or fail or analysis, unless
 * either --warning or --critical options are in use, which makes the
 * return value in some cases to match with Nagios expectations about
 * alarming. */
int main(int argc, char **argv)
{
	struct conf_t state = {
		.warning = ALARM_WARN,
		.critical = ALARM_CRIT,
		.warn_count = 0x100000000,	/* == 2^32 that is the entire IPv4 space */
		.crit_count = 0x100000000,	/* basically turns off the count criteria */
		.header_limit = 8,
		.color_mode = color_auto,
		.ranges_size = 64,
		.ip_version = IPvUNKNOWN,
		0
	};
	char output_format;
	int ret_val;

	atexit(close_stdout);
	set_program_name(argv[0]);
	prepare_memory(&state);
	set_ipv_functions(&state, IPvUNKNOWN);
	output_format = parse_command_line_opts(&state, argc, argv);

	/* Do the job */
	parse_config(&state, 1, state.dhcpdconf_file, state.shared_net_root);
	if (output_format == 'X' || output_format == 'J')
		parse_leases(&state, 1);
	else
		parse_leases(&state, 0);
	prepare_data(&state);
	do_counting(&state);
	if (state.sorts != NULL)
		mergesort_ranges(&state, state.ranges, state.num_ranges, NULL, 1);
	if (state.reverse_order == 1)
		flip_ranges(&state);
	ret_val = output_analysis(&state, output_format);
	clean_up(&state);
	return (ret_val);
}
